build/bazel/remote/output/v1/remote_output.proto

```diff
@@ -0,0 +1,295 @@
+  string output_base_id = 1;
+}
+
+message StartBuildRequest {
+  // A client-chosen value that uniquely identifies this output path.
+  // This value must be set to ensure that the remote output service is
+  // capable of managing builds concurrently.
+  //
+  // This value must be a valid filename for the operating system on
+  // which the remote output service and client are being executed. This
+  // allows the remote output service to create one subdirectory per
+  // project that needs to be built.
+  //
+  // By default, Bazel sets this value to the MD5 sum of the absolute
+  // path of the workspace directory. This is generally sufficient,
+  // though a more complex scheme may necessary in case the file system
+  // namespace is virtualized.
+  string output_base_id = 1;
+
+  // A client-chosen value that uniquely identifies this build. This
+  // value must be provided to most other methods to ensure that
+  // operations are targeted against the right output path.
+  string build_id = 2;
+
+  // The instance name that the client uses when communicating with the
+  // remote execution system. The remote output service uses this value
+  // when loading objects from the Content Addressable Storage.
+  string instance_name = 3;
+
+  // The digest function that the client uses when communicating with
+  // the remote execution system. The remote output service uses this
+  // value to ensure that FileStatus responses contain digests that were
+  // computed with right digest function.
+  build.bazel.remote.execution.v2.DigestFunction.Value digest_function = 4;
+
+  // The absolute path at which the remote output service exposes its
+  // output paths, as seen from the perspective of the client.
+  //
+  // This value needs to be provided by the client, because file system
+  // namespace virtualization may cause this directory to appear at a
+  // location that differs from the one used by the service.
+  //
+  // The purpose of this field is to ensure that the remote output
+  // service is capable of expanding symbolic links containing absolute
+  // paths.
+  string output_path_prefix = 5;
+
+  // A map of paths on the system that will become symbolic links
+  // pointing to locations inside the output path. Similar to
+  // output_path_prefix, this option is used to ensure the remote output
+  // service is capable of expanding symbolic links.
+  //
+  // Map keys are absolute paths, while map values are paths that are
+  // relative to the output path.
+  map<string, string> output_path_aliases = 6;
+}
+
+message InitialOutputPathContents {
+  // The identifier of a finalized build whose results are stored in the
+  // output path.
+  string build_id = 1;
+
+  // Paths that have been modified since the build finalized.
+  //
+  // If the remote output service freezes the contents of the output
+  // path between builds, this field can be left empty.
+  repeated string modified_paths = 2;
+}
```

Comment:
The protocol currently doesn't describe how we need to deal with remote files that disappear. A couple of options:

- We don't tell it to Bazel at the start of a build. We let the build run and somehow communicate NOT_FOUND errors. Bazel could then convert that to a LostInputsActionExecutionException. That said, I don't know whether that part of Bazel is fully functional, or whether it's something that only works well in Blaze. It looks like this requires a per-action file system, which isn't in scope for this protocol (yet). It also looks path based, which rules out the possibility of hardlinking/deduplicating identical files to a single inode at the FUSE level.
- We require that StartBuild() does a full scan of the file system and removes any files that are no longer present remotely. Any files that were absent are reported through InitialOutputPathContents.modified_paths. This means we can't do any backtracking in case files disappear during a build. This would be acceptable for Buildbarn at least, as Buildbarn's storage setup is strong enough to guarantee this in practice.

Do we think the second approach is acceptable?

Reply:
As suggested by @EricBurnett on the mailing list, the second option would be preferred. I've extended the description of StartBuild() to require that FindMissingBlobs() is called.